lib/types: Add `enumAttrs` and `enumWith`

one-d-wide opened the pull request:
Introduce `types.enumAttrs` that accepts the attribute names of the provided attribute set, and returns an option value that corresponds to the attribute value, e.g. `types.enumAttrs { "key" = "value"; ... }`, `"key"` is a valid definition, and the option value would be `"value"`.

And `types.enumWith` that performs similar convertion for the list of `{key, value}` attribute sets of any values.

###### Motivation for this change

This is useful to allow meaningful option names, that are translated to raw indexes or internal values in generated configuration.

Examples: [logLevels](https://github.com/NixOS/nixpkgs/blob/0271120180bfabe01bad1dc4163102e9204ea878/nixos/modules/services/misc/bees.nix#L9), [paths](https://github.com/NixOS/nixpkgs/blob/0271120180bfabe01bad1dc4163102e9204ea878/nixos/modules/services/networking/networkmanager.nix#L99), [attrsets](https://github.com/NixOS/nixpkgs/blob/0271120180bfabe01bad1dc4163102e9204ea878/nixos/modules/services/networking/https-dns-proxy.nix#L10).

###### Things done

- [x] Wrote type
- [x] Wrote and ran tests
- [x] Wrote docs

one-d-wide committed:
lib/types: Add `enumMap`

ncfavier commented:
What does this bring over `enum` + `apply`?

one-d-wide commented:
I see, it's even listed on [wiki](https://nixos.wiki/wiki/Declaration), btw:

> apply ('t -> any): A function used to provide an abstraction over the merged result. This function is mostly used when an usual transformation is made of the result.

Now, `enumMap` would be just a nice shortcut for such a case, which is discoverable when searching for regular `enum` in `nix repl`.

```nix
mkOption {
  type = types.enumMap enum;
  ...
}
```

one-d-wide commented:
> this can also be achieved with `coercedTo`

I don't think it could always be substituted with `types.coercedTo t f t'`, because it will leak if inputs is of type `t'` which isn't always desired  (is there just `types.apply` thing?).

> Recognizing this difference I guess `enumMap` could be improved by having the module author specify the output type.

I don't see any need for an explicit type sanitization, considering `types.enum` also allows intermixing of different types.    

There are targeted applications of `enumMap` I found in existing nixos modules (where `enum (attrNames ...)` is used now):

* Converting to indexes, allowing enumumerator and targeted type simulteneously, e.g. [logLevel](https://github.com/NixOS/nixpkgs/blob/0271120180bfabe01bad1dc4163102e9204ea878/nixos/modules/services/misc/bees.nix#L9):
  `type = either (enumMap logLevels) (attrValues logLevels);`
  Alternatively if variants aren't comprehensive:
  `type = either (enumMap logLevels) int;`

* Choosing internal values like [paths](https://github.com/NixOS/nixpkgs/blob/0271120180bfabe01bad1dc4163102e9204ea878/nixos/modules/services/networking/networkmanager.nix#L99) and [attrsets](https://github.com/NixOS/nixpkgs/blob/0271120180bfabe01bad1dc4163102e9204ea878/nixos/modules/services/networking/https-dns-proxy.nix#L10):
  `type = enumMap variants;`

However, name `enumMap` could be misleading, as it can be interpreted as affine to nix's `map`. So it should probably be changed to `enumAttrs`, emphasizing that it is an enum for attrsets.

infinisil commented:
I think the use case [you linked to](https://github.com/NixOS/nixpkgs/blob/0271120180bfabe01bad1dc4163102e9204ea878/nixos/modules/services/misc/bees.nix#L9) does make sense. It would be good to find some more examples though to really justify needing a new type for this.

If we do need it though, I would like these properties from the implementation:
- The type documentation should show:
  - All supported input values (including the ones they map to)
  - What output values they map to
  - The type of the output values
  Something like this:
  > integer or a string coerced to one, one of 0 ("foo", "bar"), 1 ("baz"), 2, 3
- The type doesn't need to support mapping non-strings for now, but its interface should be changeable to do so in the future
  This means that the interface might look like this:
  ```nix
  types.enumWith {
    # This could be added in the future when non-strings are supported
    # coercedType = string;
    finalType = int;
    # If non-strings were supported, this would require the `coercedType` to be `string`
    attributeMap = {
      foo = 0;
      bar = 0;
      baz = 1;
    };
    # This could be added in the future to support non-strings
    #map = [
    #  { input = "foo"; output = 0; }
    #  { input = "bar"; output = 0; }
    #  { input = "baz"; output = 1; }
    #];
  }
- Allow defining `enum` in terms of such an `enumWith` without breaking changes.

one-d-wide committed:
lib/types: extend `enumWith` to all values

one-d-wide committed:
lib/types: fix laziness in `enum*`

roberth commented:
I really don't like this type. It gives no access to the original value, and that could be a forward compatibility problem if you or someone ever does need the original definition value. To accommodate such a new requirement, you'd have to change the meaning of the option value, which isn't always harmless, depending on whether the option is meant to be mostly written or also read by other code.

roberth commented on nixos/doc/manual/development/option-types.section.md:
```diff
@@ -35,6 +35,18 @@ merging is handled.
+    to it's value, e.g. `types.enumAttrs { "name" = "value"; ... }`.
+    Multiple definitions cannot be merged.
+
+`types.enumWith` *`mapping`*
```
We have somewhat of a pattern where a function named `*With` takes an attribute set for its arguments, to supplement an earlier version of the function that could not be extended because of positional arguments.

I don't think we have a familiar concept of "mapping" in Nix land. "Association list" would be a more accurate term, which I think is worth to be propagated (from Haskell?). Coming to think of it, "name-value list" is even better because we have names for the pair, however, it doesn't have to be a name, right? Isn't the point of this function that it doesn't have to be a string? I think it should be "key-value list" then. Key doesn't have the same hint of stringiness.

one-d-wide commented:
> It gives no access to the original value, and that could be a forward compatibility problem

Could you please explain? Provided mapping is shown in the documentation, and can be extracted from a declared `enumWith` instance using it's functor.

> The type doesn't need to support mapping non-strings for now, but its interface should be changeable to do so in the future.

I modified `enumWith` to take a any list of input-output pairs, and also added dedicated `enumAttrs` function to accept an attrset. Is this appropriate?

>  If possible, also allow defining coercedTo in terms of such an enumWith without breaking changes.

I have implemented `enumWith` which integrates `coercedTo` logic as you described, [here](https://github.com/NixOS/nixpkgs/blob/4b95f70dd23b01d2691f9913553c35d50b4d41f4/lib/types.nix#L852).

Though I think such an approach unnecessarily overcomplicated as it can be declared using type composition and `simpleEnumWith` (that maps input-ouput pairs only, see my latest commit):

```nix 
{
  coercedType ? anything,
  coerceFunc ? null,
  finalType ? enum (catAttrs "input" mapping),
  mapping ? []
}:
  if finalType.name == "nothing" then
    simpleEnumWith mapping
  else if isNull coerceFunc then
    either finalType (simpleEnumWith mapping)
  else
    either (coercedTo coercedType coerceFunc) (simpleEnumWith mapping)
``` 

There are use cases I considered:

* one of "bar"(0), "foo"(1) 

   * `simpleEnumWith mapping` 

   * `enumWith { finalType = nothing; } mapping` 

* one of 0, 1 or one of "bar"(0), "foo"(1)

   * `either (enum (attrValues mapping)) (simpleEnumWith mapping)` 

   * `enumWith {} mapping`

* signed integer or one of "bar"(0), "foo"(1)

   * `either int (simpleEnumWith mapping)` 

   * `enumWith { finalType = int; } (simpleEnumWith mapping)`  

* (signed integer or string convertible to it) or one of "bar"(0), "foo"(1)

   * `either (coercedTo str toInt int) (simpleEnumWith mapping)`

   * `enumWith { coercedType = str; coerceFunc = toInt; finalType = int; } (simpleEnumWith mapping)`

roberth reviewed (commented):
I don't want to deny the usefulness and concision of this type. I'm biased towards writing fairly generic code that tries to be forward compatible, but a lot of modules just don't are that way.
That said, the type should come with a caveat that the original definition is lost and can not be recovered in a later change without changing the module interface in a possibly breaking manner.

one-d-wide commented on lib/types.nix:
```diff
@@ -818,34 +826,64 @@ rec {
     enum = values:
+      #assert lib.assertMsg (isList values)
+      #  "enum: `values` must be a list: ${toPretty {} values}";
+      enumWith (map (v: nameValuePair v v) (lib.unique values));
```
I see, `lib.unique` is probably particularly slow. I added it not to accidentaly trigger an assertion  (it is rather arbitrary, so I will remove it anyway)  in `enumWith`, which is off now because of some laziness tricks.

I think explicit interface assertion is much more usefull for debugging than a random exception in the implementation, so is there a way to keep it?

one-d-wide committed:
lib/types: enable asserts, rename argument, simplify functor

one-d-wide committed:
lib/types: fix laziness 2

one-d-wide committed:
lib/types: fix laziness 3

one-d-wide commented:
> That said, the type should come with a caveat that the original definition is lost and can not be recovered in a later change without changing the module interface in a possibly breaking manner.

In fact, there is a way to preserve an original value (argument passed to `enum*` function), by overriding `typeMerge` and `functor` attributes of a generated type.

```nix
enum = values:
  enumWith (map (v: keyValuePair v v) values)
  // rec {
    typeMerge = defaultTypeMerge functor;
    functor = (defaultFunctor "enum") // { type = enum; payload = values; binOp = concat; };
  };
```

Though this would block merging otherwise affine types, e.g. `enumAttrs` and `enumWith`. And I can't find a `functor` extraction used anywhere in nixpkgs, especially given that `functor` attribute is described as "internal" in `mkOption`.
